bazel: support `--rewrite` for datadriven and logic tests

**Describe the problem**

We have various datadriven and logic tests in CRDB that make use of golden datafiles, and support a `--rewrite` flag to update those golden files. Without --rewrite these datafiles are opened in [read-only mode](https://github.com/cockroachdb/datadriven/blob/e89bf9ee1861a9bf55863a1d50a3672a1ceefb18/datadriven.go#L116), so works fine. With `--rewrite` we use the [read-write mode](https://github.com/cockroachdb/datadriven/blob/e89bf9ee1861a9bf55863a1d50a3672a1ceefb18/datadriven.go#L120), which is not permitted over [data dependencies in the sandbox](https://github.com/cockroachdb/cockroach/blob/7057d2fe785535aceaf1db0c418632e305d79dc7/pkg/sql/opt/exec/execbuilder/BUILD.bazel#L57).

Strawman: we could instruct the libraries to write to another path (when run under bazel) and copy those files over those files into source tree. Bazel supports a [`TEST_UNDECLARED_OUTPUTS_DIR`](https://docs.bazel.build/versions/main/test-encyclopedia.html#initial-conditions) envvar that could work as a test-specific scratch space. It doesn't have to be `dev` copying over these files smartly, the tests could do it themselves by checking `bazel.BuiltWithBazel()` provided we include the various testdata directories as part of our `--sandbox_writable_path`. Datadriven for e.g. takes in the path it [reads/writes from](https://github.com/cockroachdb/datadriven/blob/master/datadriven.go#L114); a hacky prototype in some other project:

```go
func cp(t *testing.T, src, dest string) {
	input, err := ioutil.ReadFile(src)
	require.Nil(t, err)
	require.Nil(t, os.MkdirAll(filepath.Dir(dest), 0755))
	if err := ioutil.WriteFile(dest, input, 0644); err != nil {
		t.Fatal(err)
	}
	require.Nil(t, ioutil.WriteFile(dest, input, 0644))
}

func TestDatadriven(t *testing.T) {
	outdir := os.Getenv("TEST_UNDECLARED_OUTPUTS_DIR")
	workspace := os.Getenv("BAZEL_WORKSPACE") // passed in using --test_env=BAZEL_WORKSPACE=<...>, is there a better way?
	datadriven.Walk(t, "testdata", func(t *testing.T, path string) {
		dest := filepath.Join(outdir, path)
		cp(t, path, dest)
		defer func() { cp(t, dest, filepath.Join(repo, "<package name>", path)) }()

		datadriven.RunTest(t, dest, func(t *testing.T, d *datadriven.TestData) string {

```


**To Reproduce**

```sh
$ dev test pkg/sql/opt/exec/execbuilder -f=TestExecBuild/local/delete # read-only mode; passes
$ bazel test --config=devdarwinx86_64 //pkg/sql/opt/exec/execbuilder:execbuilder_test \  # read-write mode; fails
    --test_filter=TestExecBuild/local/delete --test_output errors --test_arg -rewrite
...
--- FAIL: TestExecBuild (0.29s)
    test_log_scope.go:73: test logs captured to: /var/folders/3j/6fxg9ckn21g5xm0z_06r7gzr0000gn/T/logTestExecBuild048473035
    test_log_scope.go:74: use -show-logs to present logs inline
    --- FAIL: TestExecBuild/local (0.23s)
        --- FAIL: TestExecBuild/local/delete (0.23s)
            logic.go:3127:
                testdata/delete:247: error while processing
            logic.go:3127: open testdata/delete: operation not permitted
    logic.go:3359: -- test log scope end --
FAIL
```

**Expected behavior**

The test to pass.

--- a/pkg/cmd/dev/test.go
+++ b/pkg/cmd/dev/test.go
@@ -12,6 +12,7 @@ package main
 
 import (
 	"fmt"
+	"path/filepath"
 	"strings"
 	"time"
 
@@ -27,6 +28,8 @@ const (
 	stressArgsFlag  = "stress-args"
 	raceFlag        = "race"
 	ignoreCacheFlag = "ignore-cache"
+	rewriteFlag     = "rewrite"
+	rewriteArgFlag  = "rewrite-arg"
 )
 
 func makeTestCmd(runE func(cmd *cobra.Command, args []string) error) *cobra.Command {
@@ -50,6 +53,8 @@ func makeTestCmd(runE func(cmd *cobra.Command, args []string) error) *cobra.Comm
 	testCmd.Flags().String(stressArgsFlag, "", "Additional arguments to pass to stress")
 	testCmd.Flags().Bool(raceFlag, false, "run tests using race builds")
 	testCmd.Flags().Bool(ignoreCacheFlag, false, "ignore cached test runs")
+	testCmd.Flags().Bool(rewriteFlag, false, "rewrite test files (only applicable for certain tests, e.g. logic and datadriven tests)")
+	testCmd.Flags().String(rewriteArgFlag, "", "additional argument to pass to -rewrite (implies --rewrite)")
 	return testCmd
 }
 
@@ -67,6 +72,8 @@ func (d *dev) test(cmd *cobra.Command, commandLine []string) error {
 	short := mustGetFlagBool(cmd, shortFlag)
 	ignoreCache := mustGetFlagBool(cmd, ignoreCacheFlag)
 	verbose := mustGetFlagBool(cmd, vFlag)
+	rewriteArg := mustGetFlagString(cmd, rewriteArgFlag)
+	rewrite := mustGetFlagBool(cmd, rewriteFlag) || (rewriteArg != "")
 
 	d.log.Printf("unit test args: stress=%t  race=%t  filter=%s  timeout=%s  ignore-cache=%t  pkgs=%s",
 		stress, race, filter, timeout, ignoreCache, pkgs)
@@ -83,6 +90,7 @@ func (d *dev) test(cmd *cobra.Command, commandLine []string) error {
 		args = append(args, "--test_sharding_strategy=disabled")
 	}
 
+	var testTargets []string
 	for _, pkg := range pkgs {
 		pkg = strings.TrimPrefix(pkg, "//")
 		pkg = strings.TrimPrefix(pkg, "./")
@@ -122,23 +130,52 @@ func (d *dev) test(cmd *cobra.Command, commandLine []string) error {
 				if err != nil {
 					return err
 				}
-				targets := strings.TrimSpace(string(out))
-				args = append(args, strings.Split(targets, "\n")...)
+				targets := strings.Split(strings.TrimSpace(string(out)), "\n")
+				testTargets = append(testTargets, targets...)
 			}
 		} else if strings.Contains(pkg, ":") {
-			args = append(args, pkg)
+			testTargets = append(testTargets, pkg)
 		} else {
 			out, err := d.exec.CommandContextSilent(ctx, "bazel", "query", fmt.Sprintf("kind(go_test, //%s:all)", pkg))
 			if err != nil {
 				return err
 			}
 			tests := strings.Split(strings.TrimSpace(string(out)), "\n")
-			args = append(args, tests...)
+			testTargets = append(testTargets, tests...)
 		}
 	}
 
+	args = append(args, testTargets...)
+
 	if ignoreCache {
 		args = append(args, "--nocache_test_results")
+	}
+	if rewrite {
+		if stress {
+			// Both of these flags require --run_under, and their usages would conflict.
+			return fmt.Errorf("cannot combine --%s and --%s", stressFlag, rewriteFlag)
+		}
+		workspace, err := d.getWorkspace(ctx)
+		if err != nil {
+			return err
+		}
+		var cdDir string
+		for _, testTarget := range testTargets {
+			dir := getDirectoryFromTarget(testTarget)
+			if cdDir != "" && cdDir != dir {
+				// We can't pass different run_under arguments for different tests
+				// in different packages.
+				return fmt.Errorf("cannot --%s for selected targets: %s. Hint: try only specifying one test target",
+					rewriteFlag, strings.Join(testTargets, ","))
+			}
+			cdDir = dir
+		}
+		args = append(args, "--run_under", fmt.Sprintf("cd %s && ", filepath.Join(workspace, cdDir)))
+		args = append(args, "--test_env=YOU_ARE_IN_THE_WORKSPACE=1")
+		args = append(args, "--test_arg", "-rewrite")
+		if rewriteArg != "" {
+			args = append(args, "--test_arg", rewriteArg)
+		}
 	}
 	if stress && timeout > 0 {
 		args = append(args, "--run_under", fmt.Sprintf("%s -maxtime=%s %s", stressTarget, timeout, stressArgs))
@@ -166,4 +203,13 @@ func (d *dev) test(cmd *cobra.Command, commandLine []string) error {
 
 	logCommand("bazel", args...)
 	return d.exec.CommandContextInheritingStdStreams(ctx, "bazel", args...)
+}
+
+func getDirectoryFromTarget(target string) string {
+	target = strings.TrimPrefix(target, "//")
+	colon := strings.LastIndex(target, ":")
+	if colon < 0 {
+		return target
+	}
+	return target[:colon]
 }

--- a/pkg/cmd/dev/testdata/recording/test.txt
+++ b/pkg/cmd/dev/testdata/recording/test.txt
@@ -167,3 +167,14 @@ bazel test //pkg/roachpb:roachpb_test //pkg/roachpb:string_test --test_output er
 
 bazel test pkg/roachpb:string_test --test_output errors
 ----
+
+bazel query 'kind(go_test, //pkg/testutils:all)'
+----
+//pkg/testutils:testutils_test
+
+bazel info workspace --color=no
+----
+go/src/github.com/cockroachdb/cockroach
+
+bazel test //pkg/testutils:testutils_test --run_under 'cd go/src/github.com/cockroachdb/cockroach/pkg/testutils && ' --test_env=YOU_ARE_IN_THE_WORKSPACE=1 --test_arg -rewrite --test_output errors
+----

--- a/pkg/cmd/dev/testdata/test.txt
+++ b/pkg/cmd/dev/testdata/test.txt
@@ -56,3 +56,9 @@ bazel test //pkg/roachpb:roachpb_test //pkg/roachpb:string_test --test_output er
 dev test pkg/roachpb:string_test
 ----
 bazel test pkg/roachpb:string_test --test_output errors
+
+dev test //pkg/testutils --rewrite
+----
+bazel query 'kind(go_test, //pkg/testutils:all)'
+bazel info workspace --color=no
+bazel test //pkg/testutils:testutils_test --run_under 'cd go/src/github.com/cockroachdb/cockroach/pkg/testutils && ' --test_env=YOU_ARE_IN_THE_WORKSPACE=1 --test_arg -rewrite --test_output errors

--- a/pkg/testutils/data_path.go
+++ b/pkg/testutils/data_path.go
@@ -11,6 +11,7 @@
 package testutils
 
 import (
+	"os"
 	"path"
 	"path/filepath"
 	"testing"
@@ -26,14 +27,15 @@ import (
 // file using TestDataPath(t, "a.txt").
 func TestDataPath(t testing.TB, relative ...string) string {
 	relative = append([]string{"testdata"}, relative...)
-	if bazel.BuiltWithBazel() {
+	// dev notifies the library that the test is running in a subdirectory of the
+	// workspace with the environment variable below.
+	if bazel.BuiltWithBazel() && os.Getenv("YOU_ARE_IN_THE_WORKSPACE") != "" {
 		runfiles, err := bazel.RunfilesPath()
 		require.NoError(t, err)
 		return path.Join(runfiles, bazel.RelativeTestTargetPath(), path.Join(relative...))
 	}
 
-	// If we're not running in Bazel, we're in the package directory and can
-	// just return a relative path.
+	// Otherwise we're in the package directory and can just return a relative path.
 	ret := path.Join(relative...)
 	ret, err := filepath.Abs(ret)
 	require.NoError(t, err)

